numpyro/distributions/discrete.py

```diff
@@ -1,171 +1,393 @@
+        batch_shape = lax.broadcast_shapes(np.shape(probs)[:-1], np.shape(total_count))
+        self.probs = promote_shapes(probs, shape=batch_shape + np.shape(probs)[-1:])[0]
+        self.total_count = promote_shapes(total_count, shape=batch_shape)[0]
+        super(Multinomial, self).__init__(batch_shape=batch_shape,
+                                          event_shape=np.shape(self.probs)[-1:],
+                                          validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return multinomial_rvs(key, self.total_count, self.probs, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        dtype = get_dtypes(self.probs)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        return gammaln(total_count + 1) + np.sum(xlogy(value, self.probs) - gammaln(value + 1), axis=-1)
+
+    @property
+    def mean(self):
+        return np.broadcast_to(self.probs * np.expand_dims(self.total_count, -1),
+                               self.batch_shape + self.event_shape)
+
+    @property
+    def variance(self):
+        return np.broadcast_to(np.expand_dims(self.total_count, -1) * self.probs * (1 - self.probs),
+                               self.batch_shape + self.event_shape)
+
+    @property
+    def support(self):
+        return constraints.multinomial(self.total_count)
+
+
+class MultinomialWithLogits(Distribution):
+    arg_constraints = {'total_count': constraints.nonnegative_integer,
+                       'logits': constraints.real}
+
+    def __init__(self, logits, total_count=1, validate_args=None):
+        if np.ndim(logits) < 1:
+            raise ValueError("`logits` parameter must be at least one-dimensional.")
+        batch_shape = lax.broadcast_shapes(np.shape(logits)[:-1], np.shape(total_count))
+        logits = logits - logsumexp(logits)
+        self.logits = promote_shapes(logits, shape=batch_shape + np.shape(logits)[-1:])[0]
+        self.total_count = promote_shapes(total_count, shape=batch_shape)[0]
+        super(MultinomialWithLogits, self).__init__(batch_shape=batch_shape,
+                                                    event_shape=np.shape(self.logits)[-1:],
+                                                    validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return multinomial_rvs(key, self.total_count, self.probs, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        dtype = get_dtypes(self.probs)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        logits = self.logits
+        logits = np.clip(logits, a_min=np.finfo(get_dtypes(logits)[0]).min)
```

Comment:
This is needed for numerical stability in `log_prob` otherwise with a value of `0`, we might end up multiplying `0 x -inf` leading to NaNs. This is to ensure that this evaluates to `0`.

Reply:
But `logits` constraint is real (not extended real) so `nan` is a good value to return in that case IMO.

Reply:
This was to ensure that the behavior is consistent with what the `xlogy` implementation does effectively, but you are right that this should be the caller's responsibility instead.